docs/adr/0003-use-queue-for-backend-services.md

```diff
@@ -0,0 +1,126 @@
+# Use queues for more resilient backend services
+
+* Status: proposed
+* Deciders: TODO
+* Date: 2023-04-06
+
+## Context and Problem Statement
+
+Monitor has had occasional service problems when traffic spikes, because of the way it and its
+external service dependencies work. Monitor relies on PostgreSQL both to store information about breaches,
+and it is also necessary for the site to function. Monitor also uses Redis as a cache layer, but if
+Postgres is down or too slow to respond, the Monitor site will start displaying error pages.
+
+* When users add an email address to Monitor, the user is registered with the breach provider (HIBP)
+  * NOTE: this uses ranges of sha256 hash sums to provide k-anonymity, so PII like raw email addresses is not sent to HIBP
+* HIBP calls Monitor's `/hibp` API for each hash range, which (for large breaches) can cause substantial database activity
+* Monitor sends email to user to notify them of breaches, which causes more users to click through to the site, causing more database activity
```

Comment:
Thanks for the review! We found out more about this as a result of investigating it together with SRE - the problem we've run into most often is that the breach data from HIBP won't be handled correctly for some reason (a bug in monitor, a change in the data format, etc) and that unhandled exception causes app servers to report as unhealthy, and k8s starts restarting things.

Since HIBP breach data comes in as a large number of separate connections, the restarts don't really resolve anything and just make the availability problem worse. We're confident that we fixed the specific bug(s) in question, but we're not as sure that we have a general solution for this problem.